File: src/switchboard/limits/bucket.py
from pathlib import Path

from redis.asyncio import Redis

_SCRIPT_SOURCE = (Path(__file__).parent / "bucket.lua").read_text()


class TokenBucket:
    def __init__(self, redis: Redis) -> None:
        self._script = redis.register_script(_SCRIPT_SOURCE)

    async def consume(
        self, key: str, capacity: int, refill_per_second: float, now: float, cost: int = 1
    ) -> tuple[bool, float]:
        allowed, remaining = await self._script(
            keys=[key], args=[capacity, refill_per_second, now, cost]
        )
        return bool(int(allowed)), float(remaining)

File: src/switchboard/limits/bucket.lua
local key = KEYS[1]
local capacity = tonumber(ARGV[1])
local refill_per_second = tonumber(ARGV[2])
local now = tonumber(ARGV[3])
local cost = tonumber(ARGV[4])

local bucket = redis.call("HMGET", key, "tokens", "updated_at")
local tokens = tonumber(bucket[1])
local updated_at = tonumber(bucket[2])

if tokens == nil then
    tokens = capacity
    updated_at = now
end

local elapsed = math.max(0, now - updated_at)
tokens = math.min(capacity, tokens + elapsed * refill_per_second)

local allowed = 0
if tokens >= cost then
    tokens = math.min(capacity, tokens - cost)
    allowed = 1
end

redis.call("HMSET", key, "tokens", tokens, "updated_at", now)
redis.call("EXPIRE", key, 3600)

return {allowed, tostring(tokens)}